File: frontend/src/types/bay.ts
/**
 * Bay Controller types — M01 Interlock Engine & Bay Controller.
 *
 * Maps to backend schemas/bay.py and routers/p3/bays.py.
 * IEC 61850 bay control: circuit breaker, disconnectors, earth switch.
 * 7 interlock rules (ILK-001..ILK-007) enforced server-side.
 */

// ── Enums ─────────────────────────────────────────────────────────────────────

export type BayType =
  | "FEEDER"
  | "BUSBAR"
  | "TIE"
  | "TRANSFORMER"
  | "MEASUREMENT"
  | "AUXILIARY";

export type BayMode = "NORMAL" | "LOCAL" | "REMOTE" | "ISOLATED" | "TEST";

export type SwitchPosition =
  | "OPEN"
  | "CLOSED"
  | "INTERMEDIATE"
  | "UNKNOWN";

export type SwitchCommand = "OPEN" | "CLOSE";

export type RelayState = "NORMAL" | "PICKUP" | "TRIP" | "LOCKOUT" | "ALARM";

// ── Synchrocheck ──────────────────────────────────────────────────────────────

export interface SynchroCheck {
  delta_voltage_percent: number;
  delta_frequency_hz: number;
  delta_phase_deg: number;
  is_in_sync: boolean;
}

// ── Bay State ─────────────────────────────────────────────────────────────────

export interface BayStateResponse {
  bay_id: string;              // UUID
  name: string;                // e.g. "BAY-01"
  display_name: string;        // e.g. "WTG Array Feeder 1"
  voltage_kv: number;          // Nominal voltage level (66 or 220 kV)
  bay_type: BayType;
  bay_mode: BayMode;
  circuit_breaker: SwitchPosition;
  disconnector_bus: SwitchPosition;
  disconnector_line: SwitchPosition;
  earth_switch: SwitchPosition;
  protection_relay: RelayState;
  manual_isolation_active: boolean;
  is_tie_cb: boolean;
  synchrocheck: SynchroCheck | null;
  last_updated: string;        // ISO-8601 UTC
}

export interface AllBaysResponse {
  bays: BayStateResponse[];
  total: number;
  energised_count: number;
  earthed_count: number;
  alarm_count: number;
}

// ── Commands ──────────────────────────────────────────────────────────────────

export interface SwitchCommandRequest {
  equipment_id: string;
  action: SwitchCommand;
  operator_id: string;
  is_auto_reclose: boolean;
  synchrocheck: SynchroCheck | null;
}

export interface CommandExecutionResponse {
  success: boolean;
  equipment_id: string;
  action: SwitchCommand;
  previous_state: SwitchPosition;
  new_state: SwitchPosition;
  message: string;
  timestamp: string;           // ISO-8601 UTC
  soe_event_id: number | null;
}

// ── Interlocks ────────────────────────────────────────────────────────────────

export interface InterlockRuleStatus {
  interlock_id: string;        // e.g. "ILK-001"
  description: string;
  currently_active: boolean;   // true = this rule is blocking
  blocking_equipment: string | null;
  blocking_state: string | null;
}

export interface InterlockStatusResponse {
  bay_id: string;
  bay_name: string;
  rules: InterlockRuleStatus[];
  all_clear: boolean;
}

// ── Validation (dry-run) ──────────────────────────────────────────────────────

export interface ValidateCommandRequest {
  bay_id: string;
  equipment_id: string;
  action: SwitchCommand;
  operator_id: string;
  is_auto_reclose: boolean;
  synchrocheck: SynchroCheck | null;
}

export interface CommandValidationResponse {
  allowed: boolean;
  blocked_by: string[];        // interlock IDs blocking the command
  reasons: string[];           // human-readable reasons
  equipment_id: string;
  action: SwitchCommand;
}

File: backend/app/schemas/bay.py
"""
Pydantic schemas for the bay controller API (M01 — Interlock Engine).

Request/response models for:
  - Bay state queries
  - Switch command execution
  - Interlock dry-run validation
  - Interlock status listing
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

# ── Shared value types ────────────────────────────────────────────


class SynchroCheckSchema(BaseModel):
    """Synchrocheck measurements for ILK-007 (tie CB close validation).

    Physics: Before closing a bus coupler to parallel two live busbars,
    both sides must be in synchronism:
      ΔV   < 5%     — voltage magnitude difference
      Δf   < 0.1 Hz — frequency difference
      Δφ   < 10°    — phase angle difference

    Reference: IEC 60909-3 synchronising conditions.
    """

    delta_voltage_percent: float = Field(
        ge=0.0,
        description="Voltage magnitude difference [%] — limit 5%",
    )
    delta_frequency_hz: float = Field(
        ge=0.0,
        description="Frequency difference [Hz] — limit 0.1 Hz",
    )
    delta_phase_deg: float = Field(
        ge=0.0,
        le=360.0,
        description="Phase angle difference [°] — limit 10°",
    )
    is_in_sync: bool = Field(
        description="True if all three conditions are within limits",
    )


# ── Bay state ─────────────────────────────────────────────────────


class BayStateResponse(BaseModel):
    """Current state of a single bay controller.

    Returned by GET /api/v1/scada/bays/{bay_id}/state
    """

    bay_id: uuid.UUID
    name: str
    display_name: str
    voltage_kv: float
    bay_type: str
    bay_mode: str = Field(description="local / remote / maintenance")
    circuit_breaker: str = Field(description="open / closed / tripped / failed / intermediate")
    disconnector_bus: str = Field(description="open / closed / intermediate")
    disconnector_line: str = Field(description="open / closed / intermediate")
    earth_switch: str = Field(description="open / closed")
    protection_relay: str = Field(description="armed / tripped / blocked / test")
    manual_isolation_active: bool
    is_tie_cb: bool
    synchrocheck: SynchroCheckSchema | None = None
    last_updated: datetime


class AllBaysResponse(BaseModel):
    """Current state of all OSS bays.

    Returned by GET /api/v1/scada/bays
    """

    bays: list[BayStateResponse]
    total: int
    energised_count: int = Field(description="Bays with CB in CLOSED state")
    earthed_count: int = Field(description="Bays with earth switch CLOSED")
    alarm_count: int = Field(description="Bays with relay TRIPPED or CB FAILED")


# ── Switch commands ───────────────────────────────────────────────


class SwitchCommandRequest(BaseModel):
    """Request body for POST /api/v1/scada/bays/{bay_id}/command.

    The command is validated through the 7-rule interlock engine before
    execution. If any interlock fires the command is rejected with 409.
    """

    equipment_id: str = Field(
        description="Equipment to operate, e.g. 'CB-STR-01' or 'ES-STR-01'",
        examples=["CB-STR-01"],
    )
    action: str = Field(
        description="Switching action: open / close / earth / unearth / rack_in / rack_out",
        examples=["close"],
    )
    operator_id: str = Field(
        description="Operator identifier for audit trail",
        examples=["operator_kaan"],
    )
    is_auto_reclose: bool = Field(
        default=False,
        description="True if command originates from auto-reclose logic (ILK-006 check)",
    )
    synchrocheck: SynchroCheckSchema | None = Field(
        default=None,
        description="Required for tie CB close commands (ILK-007 check)",
    )


class CommandExecutionResponse(BaseModel):
    """Result of a switching command execution.

    Returned by POST /api/v1/scada/bays/{bay_id}/command
    """

    success: bool
    equipment_id: str
    action: str
    previous_state: str
    new_state: str
    message: str
    timestamp: datetime
    soe_event_id: int | None = Field(
        default=None,
        description="SOE recorder event ID for this operation",
    )


# ── Interlock status ──────────────────────────────────────────────


class InterlockRuleStatus(BaseModel):
    """Status of a single interlock rule for a given bay.

    Returned as part of GET /api/v1/scada/bays/{bay_id}/interlocks
    """

    interlock_id: str = Field(description="e.g. 'ILK-001'")
    description: str = Field(description="Plain-English rule description")
    currently_active: bool = Field(description="True if this rule is currently blocking operations")
    blocking_equipment: str | None = Field(
        default=None,
        description="Equipment ID causing the interlock to fire",
    )
    blocking_state: str | None = Field(
        default=None,
        description="State of the blocking equipment",
    )


class InterlockStatusResponse(BaseModel):
    """All interlock rule statuses for a bay.

    Returned by GET /api/v1/scada/bays/{bay_id}/interlocks
    """

    bay_id: uuid.UUID
    bay_name: str
    rules: list[InterlockRuleStatus]
    all_clear: bool = Field(description="True if no interlocks currently active")


# ── Interlock dry-run ─────────────────────────────────────────────


class ValidateCommandRequest(BaseModel):
    """Request body for POST /api/v1/scada/interlocks/validate (dry-run).

    Same fields as SwitchCommandRequest but no state change is made.
    Used by the SCADA UI to colour equipment green/red before operator
    executes the real command.
    """

    bay_id: uuid.UUID
    equipment_id: str = Field(examples=["CB-STR-01"])
    action: str = Field(examples=["close"])
    operator_id: str = Field(examples=["operator_kaan"])
    is_auto_reclose: bool = False
    synchrocheck: SynchroCheckSchema | None = None


class CommandValidationResponse(BaseModel):
    """Dry-run result from POST /api/v1/scada/interlocks/validate.

    The UI shows this before the operator clicks the confirmation button:
      allowed=True  → button is green, command will succeed
      allowed=False → button is red, blocked_by shows which rules fire
    """

    allowed: bool
    blocked_by: list[str] = Field(
        description="Interlock IDs that would block this command, e.g. ['ILK-001']"
    )
    reasons: list[str] = Field(description="Human-readable block reason for each blocked_by entry")
    equipment_id: str
    action: str
